File: scenarios/moderate/cross-platform/flutter-desktop/lib/toast_notification_issues.dart
import 'package:flutter/material.dart';
class ToastNotificationIssues extends StatefulWidget {
  const ToastNotificationIssues({Key? key}) : super(key: key);
  @override
  State<ToastNotificationIssues> createState() => _ToastNotificationIssuesState();
}
class _ToastNotificationIssuesState extends State<ToastNotificationIssues> {
  final List<ToastNotification> _notifications = [];
  int _notificationCount = 0;
  @override
  void initState() {
    super.initState();
  }
    // Basic setup only
  }
  void _showToast(String type, String message) {
    setState(() {
      _notificationCount++;
      _notifications.insert(0, ToastNotification(
        id: _notificationCount,
        type: type,
        message: message,
        timestamp: DateTime.now(),
      ));
    });
    // Should implement auto-hide functionality with configurable duration
    // Basic auto-hide (5 seconds)
    Future.delayed(const Duration(seconds: 5), () {
      if (mounted) {
        _dismissToast(_notifications.first.id);
      }
    });
  }
  void _dismissToast(int id) {
    setState(() {
      _notifications.removeWhere((notification) => notification.id == id);
    });
  }
  void _clearAllNotifications() {
    setState(() {
      _notifications.clear();
      _notificationCount = 0;
    });
  }
  String _getIconForType(String type) {
    switch (type) {
      case 'success':
        return '✓';
      case 'error':
        return '✗';
      case 'warning':
        return '⚠';
      case 'info':
        return 'ℹ';
      default:
        return '•';
    }
  }
  Color _getColorForType(String type) {
    switch (type) {
      case 'success':
        return const Color(0xFF28A745);
      case 'error':
        return const Color(0xFFDC3545);
      case 'warning':
        return const Color(0xFFFFC107);
      case 'info':
        return const Color(0xFF17A2B8);
      default:
        return const Color(0xFF6C757D);
    }
  }
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color(0xFFF5F5F5),
      body: Stack(
        children: [
          SingleChildScrollView(
            padding: const EdgeInsets.all(20),
            child: Column(
              children: [
                // Main Content Card
                Card(
                  elevation: 2,
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(8),
                    side: const BorderSide(color: Color(0xFFDEE2E6)),
                  ),
                  child: Container(
                    padding: const EdgeInsets.all(30),
                    decoration: BoxDecoration(
                      color: Colors.white,
                      borderRadius: BorderRadius.circular(8),
                    ),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        // Title
                        const Text(
                          'Toast Notifications Demo',
                          style: TextStyle(
                            fontSize: 24,
                            fontWeight: FontWeight.bold,
                            color: Color(0xFF333333),
                          ),
                        ),
                        const SizedBox(height: 10),
                        // Description
                        const Text(
                          style: TextStyle(
                            fontSize: 16,
                            color: Color(0xFF666666),
                          ),
                        ),
                        const SizedBox(height: 30),
                        // Notification Controls
                        Column(
                          children: [
                            // Success Notification
                            _buildNotificationButton(
                              'Show Success Notification',
                              const Color(0xFF28A745),
                              () => _showToast('success', 'Operation completed successfully!'),
                            ),
                            const SizedBox(height: 15),
                            // Error Notification
                            _buildNotificationButton(
                              'Show Error Notification',
                              const Color(0xFFDC3545),
                              () => _showToast('error', 'An error occurred while processing your request.'),
                            ),
                            const SizedBox(height: 15),
                            // Warning Notification
                            _buildNotificationButton(
                              'Show Warning Notification',
                              const Color(0xFFFFC107),
                              () => _showToast('warning', 'Please review your input before proceeding.'),
                            ),
                            const SizedBox(height: 15),
                            // Info Notification
                            _buildNotificationButton(
                              'Show Info Notification',
                              const Color(0xFF17A2B8),
                              () => _showToast('info', 'New updates are available for download.'),
                            ),
                            const SizedBox(height: 15),
                            // Clear All Button
                            _buildNotificationButton(
                              'Clear All Notifications',
                              const Color(0xFF6C757D),
                              _clearAllNotifications,
                            ),
                          ],
                        ),
                      ],
                    ),
                  ),
                ),
                const SizedBox(height: 30),
                Card(
                  elevation: 1,
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(4),
                    side: const BorderSide(color: Color(0xFFDEE2E6)),
                  ),
                  child: Container(
                    padding: const EdgeInsets.all(20),
                    decoration: const BoxDecoration(
                      color: Color(0xFFE9ECEF),
                      borderRadius: BorderRadius.all(Radius.circular(4)),
                    ),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        const Text(
                          '• Missing dismiss button: No way to close notifications programmatically\n'
                          '• No auto-hide timer: Notifications don\'t disappear automatically\n'
                          style: TextStyle(
                            fontSize: 12,
                            color: Color(0xFF666666),
                          ),
                        ),
                        const SizedBox(height: 20),
                        // How to Fix
                        const Text(
                          'How to Fix:',
                          style: TextStyle(
                            fontSize: 16,
                            fontWeight: FontWeight.bold,
                            color: Color(0xFF333333),
                          ),
                        ),
                        const SizedBox(height: 15),
                        const Text(
                          '• Implement auto-hide timer with configurable duration\n'
                          style: TextStyle(
                            fontSize: 12,
                            color: Color(0xFF666666),
                          ),
                        ),
                      ],
                    ),
                  ),
                ),
              ],
            ),
          ),
          // Toast Notifications Container
          Positioned(
            top: 20,
            right: 20,
            child: Column(
              children: _notifications.map((notification) {
                return _buildToastNotification(notification);
              }).toList(),
            ),
          ),
        ],
      ),
    );
  }
  Widget _buildNotificationButton(String text, Color color, VoidCallback onPressed) {
    return SizedBox(
      width: double.infinity,
      child: ElevatedButton(
        onPressed: onPressed,
        style: ElevatedButton.styleFrom(
          backgroundColor: color,
          foregroundColor: color == const Color(0xFFFFC107) ? const Color(0xFF212529) : Colors.white,
          padding: const EdgeInsets.symmetric(vertical: 15, horizontal: 20),
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(4),
          ),
          elevation: 0,
        ),
        child: Text(
          text,
          style: const TextStyle(
            fontSize: 16,
            fontWeight: FontWeight.bold,
          ),
        ),
      ),
    );
  }
  Widget _buildToastNotification(ToastNotification notification) {
    final color = _getColorForType(notification.type);
    return Container(
      margin: const EdgeInsets.only(bottom: 10),
      padding: const EdgeInsets.all(15),
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(8),
        border: Border(
          left: BorderSide(color: color, width: 4),
        ),
        boxShadow: [
          BoxShadow(
            color: Colors.black.withOpacity(0.15),
            blurRadius: 12,
            offset: const Offset(0, 4),
          ),
        ],
      ),
      constraints: const BoxConstraints(
        minWidth: 300,
        maxWidth: 400,
      ),
      child: Row(
        children: [
          // Icon
          Text(
            _getIconForType(notification.type),
            style: TextStyle(
              color: color,
              fontSize: 20,
              fontWeight: FontWeight.bold,
            ),
          ),
          const SizedBox(width: 15),
          // Message
          Expanded(
            child: Text(
              notification.message,
              style: TextStyle(
                color: color,
                fontSize: 14,
                fontWeight: FontWeight.w500,
              ),
            ),
          ),
          IconButton(
            onPressed: () => _dismissToast(notification.id),
            icon: const Icon(
              Icons.close,
              color: Color(0xFF666666),
              size: 18,
            ),
            style: IconButton.styleFrom(
              backgroundColor: Colors.transparent,
              shape: const CircleBorder(),
              padding: const EdgeInsets.all(4),
            ),
          ),
        ],
      ),
    );
  }
}
class ToastNotification {
  final int id;
  final String type;
  final String message;
  final DateTime timestamp;
  ToastNotification({
    required this.id,
    required this.type,
    required this.message,
    required this.timestamp,
  });
}

File: scenarios/moderate/desktop/linux/gtk/ToastNotificationIssues.py
#!/usr/bin/env python3
"""
ToastNotificationIssues - Linux GTK Implementation
"""
import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk, GObject, GLib
import threading
import time
class ToastNotificationIssues(Gtk.Window):
    def __init__(self):
        self.set_default_size(600, 700)
        self.set_resizable(True)
        # State
        self.notifications = []
        self.notification_count = 0
        self.setup_ui()
    def setup_ui(self):
        # Main container
        main_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=20)
        main_box.set_margin_left(20)
        main_box.set_margin_right(20)
        main_box.set_margin_top(20)
        main_box.set_margin_bottom(20)
        self.add(main_box)
        # Main content card
        main_card = Gtk.Frame()
        main_card.set_shadow_type(Gtk.ShadowType.ETCHED_IN)
        main_card.set_margin_bottom(20)
        card_content = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=20)
        card_content.set_margin_left(30)
        card_content.set_margin_right(30)
        card_content.set_margin_top(30)
        card_content.set_margin_bottom(30)
        main_card.add(card_content)
        # Title
        # Description
        # Notification Controls
        controls_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=15)
        card_content.pack_start(controls_box, False, False, 0)
        # Success Notification
        self.show_success_button.set_size_request(-1, 40)
        self.show_success_button.connect("clicked", self.on_show_success_clicked)
        self.show_success_button.override_color(Gtk.StateFlags.NORMAL, Gtk.RGBA(1.0, 1.0, 1.0, 1.0))
        self.show_success_button.override_background_color(Gtk.StateFlags.NORMAL, Gtk.RGBA(0.2, 0.7, 0.2, 1.0))
        controls_box.pack_start(self.show_success_button, False, False, 0)
        # Error Notification
        self.show_error_button.set_size_request(-1, 40)
        self.show_error_button.connect("clicked", self.on_show_error_clicked)
        self.show_error_button.override_color(Gtk.StateFlags.NORMAL, Gtk.RGBA(1.0, 1.0, 1.0, 1.0))
        self.show_error_button.override_background_color(Gtk.StateFlags.NORMAL, Gtk.RGBA(0.8, 0.2, 0.2, 1.0))
        controls_box.pack_start(self.show_error_button, False, False, 0)
        # Warning Notification
        self.show_warning_button.set_size_request(-1, 40)
        self.show_warning_button.connect("clicked", self.on_show_warning_clicked)
        self.show_warning_button.override_color(Gtk.StateFlags.NORMAL, Gtk.RGBA(0.0, 0.0, 0.0, 1.0))
        self.show_warning_button.override_background_color(Gtk.StateFlags.NORMAL, Gtk.RGBA(1.0, 0.8, 0.0, 1.0))
        controls_box.pack_start(self.show_warning_button, False, False, 0)
        # Info Notification
        self.show_info_button.set_size_request(-1, 40)
        self.show_info_button.connect("clicked", self.on_show_info_clicked)
        self.show_info_button.override_color(Gtk.StateFlags.NORMAL, Gtk.RGBA(1.0, 1.0, 1.0, 1.0))
        self.show_info_button.override_background_color(Gtk.StateFlags.NORMAL, Gtk.RGBA(0.1, 0.6, 0.7, 1.0))
        controls_box.pack_start(self.show_info_button, False, False, 0)
        # Clear All Button
        self.clear_all_button.set_size_request(-1, 32)
        self.clear_all_button.connect("clicked", self.on_clear_all_clicked)
        controls_box.pack_start(self.clear_all_button, False, False, 0)
        main_box.pack_start(main_card, False, False, 0)
        # Toast Notifications Container
        self.toast_container = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=10)
        self.toast_container.set_halign(Gtk.Align.START)
        main_box.pack_start(self.toast_container, False, False, 0)
        info_frame = Gtk.Frame()
        info_frame.set_shadow_type(Gtk.ShadowType.ETCHED_IN)
        info_content = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=15)
        info_content.set_margin_left(20)
        info_content.set_margin_right(20)
        info_content.set_margin_top(20)
        info_content.set_margin_bottom(20)
        info_frame.add(info_content)
• Missing dismiss button: No way to close notifications programmatically
• No auto-hide timer: Notifications don't disappear automatically
        issues_text.set_halign(Gtk.Align.START)
        issues_text.set_line_wrap(True)
        info_content.pack_start(issues_text, False, False, 0)
        # How to Fix
        fix_title.set_markup("<b>How to Fix:</b>")
        fix_title.set_halign(Gtk.Align.START)
        info_content.pack_start(fix_title, False, False, 0)
• Implement auto-hide timer with configurable duration
        fix_text.set_halign(Gtk.Align.START)
        fix_text.set_line_wrap(True)
        info_content.pack_start(fix_text, False, False, 0)
        parent_box.pack_start(info_frame, False, False, 0)
        # Basic setup only
        self.show_success_button.set_name("show-success-button")
        self.show_error_button.set_name("show-error-button")
        self.show_warning_button.set_name("show-warning-button")
        self.show_info_button.set_name("show-info-button")
        self.clear_all_button.set_name("clear-all-button")
    def on_show_success_clicked(self, button):
        self.show_toast("success", "Operation completed successfully!")
    def on_show_error_clicked(self, button):
        self.show_toast("error", "An error occurred while processing your request.")
    def on_show_warning_clicked(self, button):
        self.show_toast("warning", "Please review your input before proceeding.")
    def on_show_info_clicked(self, button):
        self.show_toast("info", "New updates are available for download.")
    def on_clear_all_clicked(self, button):
        self.clear_all_notifications()
    def show_toast(self, notification_type, message):
        self.notification_count += 1
        notification = {
            "id": self.notification_count,
            "type": notification_type,
            "message": message
        }
        self.notifications.insert(0, notification)
        # Create toast view
        toast_view = self.create_toast_view(notification)
        self.toast_container.pack_start(toast_view, False, False, 0)
        # Should implement auto-hide functionality with configurable duration
        # Basic auto-hide (5 seconds)
        def auto_hide():
            time.sleep(5)
            GLib.idle_add(self.remove_toast, notification)
        threading.Thread(target=auto_hide, daemon=True).start()
    def create_toast_view(self, notification):
        # Container
        container = Gtk.Frame()
        container.set_shadow_type(Gtk.ShadowType.OUT)
        container.set_name(f"toast-{notification['id']}")
        # Content box
        content_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=10)
        content_box.set_margin_left(15)
        content_box.set_margin_right(15)
        content_box.set_margin_top(15)
        content_box.set_margin_bottom(15)
        container.add(content_box)
        # Icon
        # Message
        dismiss_button.set_size_request(24, 24)
        dismiss_button.set_relief(Gtk.ReliefStyle.NONE)
        dismiss_button.connect("clicked", self.on_dismiss_toast, notification['id'])
        content_box.pack_start(dismiss_button, False, False, 0)
        # Set background color based on type
        container.override_background_color(Gtk.StateFlags.NORMAL, self.get_background_color_for_type(notification['type']))
        # Basic setup only
        container.set_name(f"toast-{notification['id']}")
        return container
    def get_icon_for_type(self, notification_type):
        icons = {
            "success": "✓",
            "error": "✗",
            "warning": "⚠",
            "info": "ℹ"
        }
        return icons.get(notification_type, "•")
    def get_color_for_type(self, notification_type):
        colors = {
            "success": Gtk.RGBA(0.2, 0.7, 0.2, 1.0),
            "error": Gtk.RGBA(0.8, 0.2, 0.2, 1.0),
            "warning": Gtk.RGBA(0.8, 0.6, 0.0, 1.0),
            "info": Gtk.RGBA(0.1, 0.6, 0.7, 1.0)
        }
        return colors.get(notification_type, Gtk.RGBA(0.4, 0.4, 0.4, 1.0))
    def get_background_color_for_type(self, notification_type):
        colors = {
            "success": Gtk.RGBA(0.2, 0.7, 0.2, 0.1),
            "error": Gtk.RGBA(0.8, 0.2, 0.2, 0.1),
            "warning": Gtk.RGBA(1.0, 0.8, 0.0, 0.1),
            "info": Gtk.RGBA(0.1, 0.6, 0.7, 0.1)
        }
        return colors.get(notification_type, Gtk.RGBA(0.9, 0.9, 0.9, 1.0))
    def on_dismiss_toast(self, button, notification_id):
        self.remove_toast_by_id(notification_id)
    def remove_toast(self, notification):
        self.remove_toast_by_id(notification['id'])
    def remove_toast_by_id(self, notification_id):
        # Remove from notifications list
        self.notifications = [n for n in self.notifications if n['id'] != notification_id]
        # Remove from UI
        for child in self.toast_container.get_children():
            if child.get_name() == f"toast-{notification_id}":
                self.toast_container.remove(child)
                break
    def clear_all_notifications(self):
        self.notifications.clear()
        self.notification_count = 0
        # Remove all toast views
        for child in self.toast_container.get_children():
            self.toast_container.remove(child)
    # Should implement methods to:
    # 4. Provide programmatic access to notification content
def main():
    app = Gtk.Application()
    app.connect('activate', lambda app: ToastNotificationIssues().show_all())
    app.run()
if __name__ == '__main__':
    main()
